src/client/src/tscServer.c

```diff
@@ -51,40 +52,92 @@ static void tscSetDnodeIpList(SSqlObj* pSql, SCMVgroupInfo* pVgroupInfo) {
     pIpList->numOfIps = 0;
     return;
   }
-  
+
   pIpList->numOfIps = pVgroupInfo->numOfIps;
   for(int32_t i = 0; i < pVgroupInfo->numOfIps; ++i) {
     strcpy(pIpList->fqdn[i], pVgroupInfo->ipAddr[i].fqdn);
     pIpList->port[i] = pVgroupInfo->ipAddr[i].port;
   }
 }
+void tscIpSetCopy(SRpcIpSet *dst, SRpcIpSet *src) {
+  dst->numOfIps = src->numOfIps;
+  dst->inUse = src->inUse;
+  for (int32_t i = 0; i < src->numOfIps; ++i) {
+     dst->port[i] = src->port[i];
+     strncpy(dst->fqdn[i], src->fqdn[i], TSDB_FQDN_LEN);
+  }
+}
+static void tscDumpMgmtIpSet(SRpcIpSet *ipSet) {
+  taosCorBeginRead(&tscMgmtIpSet.version);
+  SRpcIpSet*  src = &tscMgmtIpSet.ipSet;
+  tscIpSetCopy(ipSet, src); 
+  taosCorEndRead(&tscMgmtIpSet.version);
+}  
+static void tscIpSetHtons(SRpcIpSet *s) {
+   for (int32_t i = 0; i < s->numOfIps; i++) {
+      s->port[i] = htons(s->port[i]);    
+   }
+} 
+bool tscIpSetIsEqual(SRpcIpSet *s1, SRpcIpSet *s2) {
+   if (s1->numOfIps != s2->numOfIps || s1->inUse != s2->inUse) {
+     return false;
+   } 
+   for (int32_t i = 0; i < s1->numOfIps; i++) {
+     if (s1->port[i] != s2->port[i] 
+        || strncmp(s1->fqdn[i], s2->fqdn[i], TSDB_FQDN_LEN) != 0)
+        return false;
+   }
+   return true;
+}
+void tscUpdateMgmtIpList(SRpcIpSet *pIpSet) {
+  // no need to update if equal
+  taosCorBeginWrite(&tscMgmtIpSet.version);
+  // or copy directly,  tscMgmtIpSet.ipSet = *pIpSet
+  SRpcIpSet *mgmtIpSet = &tscMgmtIpSet.ipSet;
+  tscIpSetCopy(mgmtIpSet, pIpSet);
+  taosCorEndWrite(&tscMgmtIpSet.version);
+}
+static void tscDumpIpSetFromVgroupInfo(SCMCorVgroupInfo *pVgroupInfo, SRpcIpSet *pIpSet) {
+  if (pVgroupInfo == NULL) { return;}
+  taosCorBeginRead(&pVgroupInfo->version);
+  int8_t inUse = pVgroupInfo->inUse;
+  pIpSet->inUse = (inUse >= 0 && inUse < TSDB_MAX_REPLICA) ? inUse: 0; 
+  pIpSet->numOfIps = pVgroupInfo->numOfIps;  
+  for (int32_t i = 0; i < pVgroupInfo->numOfIps; ++i) {
+    strncpy(pIpSet->fqdn[i], pVgroupInfo->ipAddr[i].fqdn, TSDB_FQDN_LEN);
+    pIpSet->port[i] = pVgroupInfo->ipAddr[i].port;
+  }
+  taosCorEndRead(&pVgroupInfo->version);
+}
+
+static void tscUpdateVgroupInfo(SSqlObj *pObj, SRpcIpSet *pIpSet) {
+  SSqlCmd *pCmd = &pObj->cmd;
+  STableMetaInfo *pTableMetaInfo = tscGetTableMetaInfoFromCmd(pCmd, pCmd->clauseIndex, 0);
+  if (pTableMetaInfo == NULL || pTableMetaInfo->pTableMeta == NULL) { return;}
+  SCMCorVgroupInfo *pVgroupInfo = &pTableMetaInfo->pTableMeta->corVgroupInfo;
 
+  taosCorBeginWrite(&pVgroupInfo->version);
+  //TODO(dengyihao), dont care vgid 
+  pVgroupInfo->inUse = pIpSet->inUse;
+  pVgroupInfo->numOfIps = pIpSet->numOfIps;
+  for (int32_t i = 0; pVgroupInfo->numOfIps; i++) {
+    strncpy(pVgroupInfo->ipAddr[i].fqdn, pIpSet->fqdn[i], TSDB_FQDN_LEN);
+    pVgroupInfo->ipAddr[i].port = pIpSet->port[i];
+  }
+  taosCorEndWrite(&pVgroupInfo->version);
```

Comment:
rename this function to taosEndCopyOnWrite

Reply:
the new name is not appropriate as  this is not "copy on write"